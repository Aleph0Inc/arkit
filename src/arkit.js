"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const logger_1 = require("./logger");
const config_1 = require("./config");
const parser_1 = require("./parser");
const generator_1 = require("./generator");
exports.arkit = (directory) => {
    const config = new config_1.Config(directory);
    logger_1.debug(`Config`);
    logger_1.trace(config);
    const parser = new parser_1.Parser(config);
    const files = parser.parse();
    logger_1.trace('Parsed files');
    logger_1.trace(files);
    const generator = new generator_1.Generator(config, files);
    return Promise.all(config.outputs.map(output => {
        const puml = generator.generatePlantUML(output);
        if (output.path) {
            for (const outputPath of config.array(output.path)) {
                const fullExportPath = path.join(config.directory, outputPath);
                const ext = path.extname(fullExportPath);
                if (fs.existsSync(fullExportPath)) {
                    fs.unlinkSync(fullExportPath);
                }
                if (ext === '.puml') {
                    fs.writeFileSync(fullExportPath, puml);
                }
                if (ext === '.svg') {
                    generator.convertToSVG(puml).then(svg => {
                        fs.writeFileSync(fullExportPath, svg);
                    }).catch(err => {
                        throw err;
                    });
                }
            }
            return puml;
        }
        return generator.convertToSVG(puml).then(svg => {
            console.log(svg);
            return puml;
        }).catch(err => {
            throw err;
        });
    }));
};
