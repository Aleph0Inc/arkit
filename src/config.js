"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const logger_1 = require("./logger");
/**
 * Component name formats
 */
var ComponentNameFormat;
(function (ComponentNameFormat) {
    ComponentNameFormat["BASE_NAME"] = "base";
    ComponentNameFormat["FULL_NAME"] = "full";
    ComponentNameFormat["COMPLETE_PATH"] = "complete";
})(ComponentNameFormat = exports.ComponentNameFormat || (exports.ComponentNameFormat = {}));
var OutputDirection;
(function (OutputDirection) {
    OutputDirection["HORIZONTAL"] = "horizontal";
    OutputDirection["VERTICAL"] = "vertical";
})(OutputDirection = exports.OutputDirection || (exports.OutputDirection = {}));
const DEFAULT_COMPONENTS = {
    type: 'Component',
    patterns: ['**/*.ts', '**/*.js', '**/*.jsx', '**/*.tsx']
};
exports.DEFAULT_CONFIG = {
    components: DEFAULT_COMPONENTS,
    excludePatterns: ['node_modules/**', 'test/**', '**/*.test.*', '**/*.spec.*'],
    output: {}
};
class Config {
    constructor(directory) {
        this.directory = directory;
        this.patterns = [];
        this.excludePatterns = [];
        this.extensions = ['.js', '.ts', '.jsx', '.tsx'];
        logger_1.info('Working directory', directory);
        const userConfigPath = path.join(this.directory, 'arkit');
        const userConfig = this.safeRequire(userConfigPath);
        this.components = this.array(userConfig && userConfig.components) || [];
        if (!this.components.length) {
            this.components.push(DEFAULT_COMPONENTS);
        }
        this.outputs = this.array(userConfig && userConfig.output || exports.DEFAULT_CONFIG.output);
        if (userConfig && userConfig.excludePatterns) {
            this.excludePatterns.push(...userConfig.excludePatterns);
        }
        else if (!userConfig && exports.DEFAULT_CONFIG.excludePatterns) {
            this.excludePatterns.push(...exports.DEFAULT_CONFIG.excludePatterns);
        }
        for (const component of this.components) {
            if (component.patterns) {
                this.patterns.push(...component.patterns);
            }
            if (component.excludePatterns) {
                this.excludePatterns.push(...component.excludePatterns);
            }
        }
    }
    safeRequire(path) {
        try {
            return require(path);
        }
        catch (e) {
            logger_1.warn(e.toString());
        }
    }
    array(input) {
        if (input) {
            return [].concat(input);
        }
    }
}
exports.Config = Config;
