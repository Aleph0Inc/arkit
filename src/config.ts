import * as path from 'path'
import { ComponentSchema, ConfigSchema, Options, OutputSchema } from './schema'
import { array, safeRequire } from './utils'

const DEFAULT_COMPONENTS: ComponentSchema[] = [
  {
    type: 'Component',
    patterns: ['**/*.ts', '**/*.js', '**/*.jsx', '**/*.tsx']
  },
  {
    type: 'Dependency',
    patterns: ['node_modules/*']
  }
]

export class Config {
  directory: string
  components: ComponentSchema[]
  outputs: OutputSchema[]
  patterns: string[] = []
  excludePatterns: string[]
  extensions = ['.js', '.ts', '.jsx', '.tsx']

  constructor (options: Options) {
    this.directory = options.directory
    const userConfigPath = path.resolve(this.directory, 'arkit')
    const userConfig = safeRequire<ConfigSchema>(userConfigPath)

    this.components = array(userConfig && userConfig.components) || []

    if (!this.components.length) {
      this.components.push(...DEFAULT_COMPONENTS)
    }

    this.outputs = this.getOutputs(options, userConfig)
    this.excludePatterns = this.getExcludePatterns(options, userConfig)

    for (const component of this.components) {
      if (component.patterns) {
        this.patterns.push(...component.patterns)
      }
    }
  }

  private getOutputs (options: Options, userConfig?: ConfigSchema): OutputSchema[] {
    const userConfigOutput = userConfig && userConfig.output
    const outputOption = options.output && options.output.length ? options.output : undefined
    const firstOption = options.first && options.first.length ? options.first : undefined
    const shouldGenerateOutput = outputOption || firstOption || !userConfigOutput

    if (!shouldGenerateOutput) {
      return array(userConfigOutput)!
    }

    const firstGroup = firstOption ? [{
      first: true,
      patterns: firstOption
    }] : []

    return [
      {
        path: outputOption,
        groups: [
          ...firstGroup,
          { type: 'Dependencies', components: ['Dependency'] },
          {} // everything else
        ]
      }
    ]
  }

  private getExcludePatterns (options: Options, userConfig?: ConfigSchema): string[] {
    const excludePatterns: string[] = []

    if (options.exclude) {
      excludePatterns.push(...options.exclude)
    }

    if (userConfig && userConfig.excludePatterns) {
      excludePatterns.push(...userConfig.excludePatterns)
    }

    for (const component of this.components) {
      if (component.excludePatterns) {
        excludePatterns.push(...component.excludePatterns)
      }
    }

    return excludePatterns
  }
}
