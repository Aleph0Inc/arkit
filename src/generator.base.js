"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("./schema");
const logger_1 = require("./logger");
const path = require("path");
const nanomatch = require("nanomatch");
exports.EMPTY_LAYER = Symbol('__empty_layer__');
var Context;
(function (Context) {
    Context[Context["LAYER"] = 0] = "LAYER";
    Context[Context["RELATIONSHIP"] = 1] = "RELATIONSHIP";
})(Context = exports.Context || (exports.Context = {}));
class GeneratorBase {
    constructor(config, files) {
        this.config = config;
        this.files = files;
    }
    generateComponents(output) {
        const components = Object.keys(this.files).reduce((components, filename) => {
            const schema = this.findComponentSchema(output, filename);
            if (schema) {
                const name = this.getComponentName(filename, schema);
                components.set(filename, {
                    name,
                    filename,
                    isImported: false,
                    type: schema.type,
                    layer: exports.EMPTY_LAYER,
                    imports: Object.keys(this.files[filename].imports)
                });
            }
            return components;
        }, new Map());
        for (const component of components.values()) {
            for (const potentialComponent of components.values()) {
                if (potentialComponent.imports.includes(component.filename)) {
                    component.isImported = true;
                    break;
                }
            }
        }
        return components;
    }
    generateLayers(output, allComponents) {
        const groups = this.config.array(output.groups) || [{}];
        const ungroupedComponents = new Map(allComponents);
        const grouppedComponents = new Map();
        const layers = new Map();
        groups.forEach(group => {
            const layerType = group.type || exports.EMPTY_LAYER;
            if (!layers.has(layerType)) {
                layers.set(layerType, new Set());
            }
            Array.from(ungroupedComponents.entries())
                .filter(([filename, component]) => {
                return this.verifyComponentFilters(group, component);
            })
                .forEach(([filename, component]) => {
                component.layer = layerType;
                component.first = group.first;
                component.last = group.last;
                layers.get(layerType).add(component);
                grouppedComponents.set(component.filename, component);
                ungroupedComponents.delete(filename);
                return component;
            });
        });
        if (ungroupedComponents.size) {
            logger_1.trace('Ungrouped components');
            logger_1.trace(Array.from(ungroupedComponents.values()));
        }
        const filenamesFromFirstComponents = new Set();
        for (const component of grouppedComponents.values()) {
            if (component.first) {
                this.collectImportedFilenames(component, grouppedComponents, filenamesFromFirstComponents);
            }
        }
        if (filenamesFromFirstComponents.size) {
            logger_1.trace('Filenames from first components');
            logger_1.trace(Array.from(filenamesFromFirstComponents));
            for (const [filename, component] of allComponents) {
                if (!filenamesFromFirstComponents.has(filename)) {
                    for (const components of layers.values()) {
                        components.delete(component);
                    }
                    ungroupedComponents.delete(filename);
                    allComponents.delete(filename);
                }
            }
        }
        if (ungroupedComponents.size) {
            logger_1.trace('Ungrouped components leftovers');
            logger_1.trace(Array.from(ungroupedComponents.values()));
        }
        return layers;
    }
    collectImportedFilenames(component, components, filenames) {
        if (filenames.has(component.filename))
            return;
        filenames.add(component.filename);
        if (!component.last) {
            component.imports.forEach(importedFilename => {
                const importedComponent = components.get(importedFilename);
                if (importedComponent) {
                    this.collectImportedFilenames(importedComponent, components, filenames);
                }
            });
        }
        else {
            component.imports = [];
        }
    }
    resolveConflictingComponentNames(components) {
        const componentsByName = {};
        for (const component of components.values()) {
            componentsByName[component.name] = componentsByName[component.name] || [];
            componentsByName[component.name].push(component);
        }
        for (const name in componentsByName) {
            const components = componentsByName[name];
            const differentFilenames = new Set(components.map(component => component.filename));
            const shouldPrefixWithDirectory = differentFilenames.size > 1 || name === 'index';
            if (shouldPrefixWithDirectory) {
                for (const component of components) {
                    const dir = path.basename(path.dirname(component.filename));
                    component.name = path.join(dir, component.name);
                }
            }
        }
        return components;
    }
    sortComponentsByName(components) {
        const sortedComponents = new Map(Array.from(components.entries()).sort((a, b) => a[1].name.localeCompare(b[1].name)));
        for (const component of components.values()) {
            component.imports = component.imports
                .filter(importedFilename => components.has(importedFilename))
                .sort((a, b) => {
                const componentA = components.get(a);
                const componentB = components.get(b);
                return componentA.name.localeCompare(componentB.name);
            });
        }
        return sortedComponents;
    }
    findComponentSchema(output, filename) {
        const componentSchema = this.config.components.find(componentSchema => {
            const outputFilters = [output, ...this.config.array(output.groups) || []];
            const includedInOutput = outputFilters.some(outputFilter => this.verifyComponentFilters(outputFilter, componentSchema));
            if (includedInOutput) {
                return !!componentSchema.patterns && nanomatch.some(filename, componentSchema.patterns);
            }
            else {
                return false;
            }
        });
        if (!componentSchema) {
            throw new Error(`Component schema not found: ${filename}`);
        }
        return componentSchema;
    }
    verifyComponentFilters(filters, component) {
        const matchesPatterns = !('filename' in component) || !filters.patterns || nanomatch.some(component.filename, filters.patterns);
        const matchesComponents = !filters.components || filters.components.some(type => type === component.type);
        return matchesPatterns && matchesComponents;
    }
    getComponentName(filename, componentConfig) {
        const nameFormat = componentConfig.format;
        const a = 1;
        if (nameFormat === schema_1.ComponentNameFormat.FULL_NAME) {
            return path.basename(filename);
        }
        return path.basename(filename, path.extname(filename));
    }
    getAllComponents(layers) {
        return []
            .concat(...[...layers.values()].map(components => [...components]));
    }
}
exports.GeneratorBase = GeneratorBase;
