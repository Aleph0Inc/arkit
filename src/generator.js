"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const https = require("https");
const nanomatch = require("nanomatch");
const config_1 = require("./config");
const logger_1 = require("./logger");
const EMPTY_LAYER = Symbol('__empty_layer__');
class Generator {
    constructor(config, files) {
        this.config = config;
        this.files = files;
    }
    generateComponents(output) {
        const components = Object.keys(this.files).reduce((components, filename) => {
            const schema = this.findComponentSchema(output, filename);
            if (schema) {
                const name = this.getComponentName(filename, schema);
                components.set(filename, {
                    name,
                    filename,
                    isImported: false,
                    type: schema.type,
                    layer: EMPTY_LAYER,
                    imports: Object.keys(this.files[filename].imports)
                });
            }
            return components;
        }, new Map());
        for (const component of components.values()) {
            for (const potentialComponent of components.values()) {
                if (potentialComponent.imports.includes(component.filename)) {
                    component.isImported = true;
                    break;
                }
            }
        }
        return components;
    }
    generateLayers(output, allComponents) {
        const groups = this.config.array(output.groups) || [{}];
        const ungroupedComponents = new Map(allComponents);
        const grouppedComponents = new Map();
        const layers = new Map();
        groups.forEach(group => {
            const layerType = group.type || EMPTY_LAYER;
            if (!layers.has(layerType)) {
                layers.set(layerType, new Set());
            }
            Array.from(ungroupedComponents.entries())
                .filter(([filename, component]) => {
                return this.verifyComponentFilters(group, component);
            })
                .forEach(([filename, component]) => {
                component.layer = layerType;
                component.first = group.first;
                component.last = group.last;
                layers.get(layerType).add(component);
                grouppedComponents.set(component.filename, component);
                ungroupedComponents.delete(filename);
                return component;
            });
        });
        if (ungroupedComponents.size) {
            logger_1.trace('Ungrouped components');
            logger_1.trace(Array.from(ungroupedComponents.values()));
        }
        const filenamesFromFirstComponents = new Set();
        for (const component of grouppedComponents.values()) {
            if (component.first) {
                this.collectImportedFilenames(component, grouppedComponents, filenamesFromFirstComponents);
            }
        }
        if (filenamesFromFirstComponents.size) {
            logger_1.trace('Filenames from first components');
            logger_1.trace(Array.from(filenamesFromFirstComponents));
            for (const [filename, component] of allComponents) {
                if (!filenamesFromFirstComponents.has(filename)) {
                    for (const components of layers.values()) {
                        components.delete(component);
                    }
                    ungroupedComponents.delete(filename);
                    allComponents.delete(filename);
                }
            }
        }
        if (ungroupedComponents.size) {
            logger_1.trace('Ungrouped components leftovers');
            logger_1.trace(Array.from(ungroupedComponents.values()));
        }
        return layers;
    }
    collectImportedFilenames(component, components, filenames) {
        if (filenames.has(component.filename))
            return;
        filenames.add(component.filename);
        if (!component.last) {
            component.imports.forEach(importedFilename => {
                const importedComponent = components.get(importedFilename);
                if (importedComponent) {
                    this.collectImportedFilenames(importedComponent, components, filenames);
                }
            });
        }
        else {
            component.imports = [];
        }
    }
    resolveConflictingComponentNames(components) {
        const componentsByName = {};
        for (const component of components.values()) {
            componentsByName[component.name] = componentsByName[component.name] || [];
            componentsByName[component.name].push(component);
        }
        for (const name in componentsByName) {
            const components = componentsByName[name];
            const differentFilenames = new Set(components.map(component => component.filename));
            const shouldPrefixWithDirectory = differentFilenames.size > 1 || name === 'index';
            if (shouldPrefixWithDirectory) {
                for (const component of components) {
                    const dir = path.basename(path.dirname(component.filename));
                    component.name = path.join(dir, component.name);
                }
            }
        }
        return components;
    }
    sortComponentsByName(components) {
        const sortedComponents = new Map(Array.from(components.entries()).sort((a, b) => a[1].name.localeCompare(b[1].name)));
        for (const component of components.values()) {
            component.imports = component.imports
                .filter(importedFilename => components.has(importedFilename))
                .sort((a, b) => {
                const componentA = components.get(a);
                const componentB = components.get(b);
                return componentA.name.localeCompare(componentB.name);
            });
        }
        return sortedComponents;
    }
    findComponentSchema(output, filename) {
        const componentSchema = this.config.components.find(componentSchema => {
            const outputFilters = [output, ...this.config.array(output.groups) || []];
            const includedInOutput = outputFilters.some(outputFilter => this.verifyComponentFilters(outputFilter, componentSchema));
            if (includedInOutput) {
                return !!componentSchema.patterns && nanomatch.some(filename, componentSchema.patterns);
            }
            else {
                return false;
            }
        });
        if (!componentSchema) {
            throw new Error(`Component schema not found: ${filename}`);
        }
        return componentSchema;
    }
    verifyComponentFilters(filters, component) {
        const matchesPatterns = !('filename' in component) || !filters.patterns || nanomatch.some(component.filename, filters.patterns);
        const matchesComponents = !filters.components || filters.components.some(type => type === component.type);
        return matchesPatterns && matchesComponents;
    }
    getComponentName(filename, componentConfig) {
        const nameFormat = componentConfig.format;
        if (nameFormat === config_1.ComponentNameFormat.FULL_NAME) {
            return path.basename(filename);
        }
        return path.basename(filename, path.extname(filename));
    }
    match(filename, pattern) {
        return nanomatch(filename, pattern);
    }
    generatePlantUML(output) {
        logger_1.debug('Generating components...');
        const components = this.sortComponentsByName(this.resolveConflictingComponentNames(this.generateComponents(output)));
        logger_1.trace(Array.from(components.values()));
        logger_1.debug('Generating layers...');
        const layers = this.generateLayers(output, components);
        logger_1.trace(Array.from(layers.keys()));
        const puml = ['@startuml'];
        puml.push(this.generatePlantUMLSkin(output, layers));
        for (const [layer, components] of layers.entries()) {
            puml.push(this.generatePlantUMLLayer(layer, components));
        }
        puml.push(this.generatePlantUMLRelationships(layers));
        puml.push('');
        puml.push('@enduml');
        return puml.join('\n');
    }
    convertToSVG(puml) {
        return new Promise((resolve, reject) => {
            const req = https.request('https://arkit.herokuapp.com/svg', {
                method: 'post',
                headers: {
                    'Content-Type': 'text/plain',
                    'Content-Length': puml.length
                }
            }, res => {
                let svg = [''];
                res.on('data', data => svg.push(data));
                res.on('end', () => resolve(svg.join('')));
            }).on('error', reject);
            req.write(puml);
            req.end();
        });
    }
    generatePlantUMLLayer(layer, components) {
        if (!components.size)
            return '';
        const puml = [''];
        const isLayer = typeof layer === 'string';
        if (isLayer)
            puml.push(`package "${layer}" {`);
        for (const component of components) {
            const componentPuml = [this.generatePlantUMLComponent(component)];
            if (isLayer)
                componentPuml.unshift('  ');
            puml.push(componentPuml.join(''));
        }
        if (isLayer)
            puml.push('}');
        return puml.join('\n');
    }
    generatePlantUMLComponent(component) {
        const puml = [];
        puml.push('(');
        if (!component.isImported)
            puml.push('<b>');
        puml.push(component.name);
        if (!component.isImported)
            puml.push('</b>');
        puml.push(')');
        return puml.join('');
    }
    generatePlantUMLRelationships(layers) {
        const puml = [''];
        const components = this.getAllComponents(layers)
            .sort((a, b) => a.name.localeCompare(b.name));
        for (const component of components) {
            for (const importedFilename of component.imports) {
                const importedComponent = components.find(importedComponent => importedComponent.filename === importedFilename);
                if (!importedComponent)
                    continue;
                const numberOfLevels = path.dirname(path.relative(component.filename, importedFilename)).split(path.sep).length;
                const connectionLength = Math.max(1, Math.min(4, numberOfLevels));
                const connectionSign = !component.isImported ? '=' : component.layer === importedComponent.layer && typeof component.layer === 'string' ? '~' : '-';
                const connection = connectionSign.repeat(connectionLength) + '>';
                puml.push([
                    this.generatePlantUMLComponent(component),
                    connection,
                    this.generatePlantUMLComponent(importedComponent)
                ].join(' '));
            }
        }
        return puml.join('\n');
    }
    /**
     * https://github.com/plantuml/plantuml/blob/master/src/net/sourceforge/plantuml/SkinParam.java
     */
    generatePlantUMLSkin(output, layers) {
        const puml = [''];
        puml.push('scale max 1200 width');
        const direction = output.direction || this.getAllComponents(layers).length > 20 ? config_1.OutputDirection.HORIZONTAL : config_1.OutputDirection.VERTICAL;
        if (direction === config_1.OutputDirection.HORIZONTAL) {
            puml.push('left to right direction');
        }
        else {
            puml.push('top to bottom direction');
        }
        puml.push('skinparam monochrome true');
        puml.push('skinparam shadowing false');
        puml.push('skinparam nodesep 22');
        puml.push('skinparam defaultFontName Tahoma');
        puml.push('skinparam defaultFontSize 14');
        puml.push(`
'oval
skinparam usecase {
  borderThickness 1
}
    `);
        return puml.join('\n');
    }
    getAllComponents(layers) {
        return []
            .concat(...[...layers.values()].map(components => [...components]));
    }
}
exports.Generator = Generator;
