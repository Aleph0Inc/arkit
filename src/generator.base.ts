import { Config } from './config'
import { ComponentFilters, ComponentNameFormat, ComponentSchema, OutputSchema } from './schema'
import { trace } from './logger'
import * as path from 'path'
import { Files } from './parser'
import * as nanomatch from 'nanomatch'

export const EMPTY_LAYER = Symbol('__empty_layer__')

export interface Component {
  name: string
  type: string
  filename: string
  imports: string[]
  layer: string | typeof EMPTY_LAYER
  isImported: boolean
  first?: boolean
  last?: boolean
}

export interface Components extends Map<string, Component> {
}

export interface Layers extends Map<string | typeof EMPTY_LAYER, Set<Component>> {
}

export enum Context {
  LAYER,
  RELATIONSHIP
}

export class GeneratorBase {
  protected config: Config
  private files: Files

  constructor (
    config: Config,
    files: Files
  ) {
    this.config = config
    this.files = files
  }

  protected generateComponents (output: OutputSchema): Components {
    const components = Object.keys(this.files).reduce((components, filename) => {
      const schema = this.findComponentSchema(output, filename)

      if (schema) {
        const name = this.getComponentName(filename, schema)

        components.set(filename, {
          name,
          filename,
          isImported: false,
          type: schema.type,
          layer: EMPTY_LAYER,
          imports: Object.keys(this.files[filename].imports)
        })
      }

      return components
    }, new Map() as Components)

    for (const component of components.values()) {
      for (const potentialComponent of components.values()) {
        if (potentialComponent.imports.includes(component.filename)) {
          component.isImported = true
          break
        }
      }
    }

    return components
  }

  protected generateLayers (output: OutputSchema, allComponents: Components): Layers {
    const groups = this.config.array(output.groups) || [{}]
    const ungroupedComponents: Components = new Map(allComponents)
    const grouppedComponents = new Map<string, Component>()
    const layers: Layers = new Map()

    groups.forEach(group => {
      const layerType = group.type || EMPTY_LAYER

      if (!layers.has(layerType)) {
        layers.set(layerType, new Set())
      }

      Array.from(ungroupedComponents.entries())
        .filter(([filename, component]) => {
          return this.verifyComponentFilters(group, component)
        })
        .forEach(([filename, component]) => {
          component.layer = layerType
          component.first = group.first
          component.last = group.last
          layers.get(layerType)!.add(component)
          grouppedComponents.set(component.filename, component)
          ungroupedComponents.delete(filename)
          return component
        })
    })

    if (ungroupedComponents.size) {
      trace('Ungrouped components')
      trace(Array.from(ungroupedComponents.values()))
    }

    const filenamesFromFirstComponents = new Set<string>()

    for (const component of grouppedComponents.values()) {
      if (component.first) {
        this.collectImportedFilenames(component, grouppedComponents, filenamesFromFirstComponents)
      }
    }

    if (filenamesFromFirstComponents.size) {
      trace('Filenames from first components')
      trace(Array.from(filenamesFromFirstComponents))

      for (const [filename, component] of allComponents) {
        if (!filenamesFromFirstComponents.has(filename)) {
          for (const components of layers.values()) {
            components.delete(component)
          }

          ungroupedComponents.delete(filename)
          allComponents.delete(filename)
        }
      }
    }

    if (ungroupedComponents.size) {
      trace('Ungrouped components leftovers')
      trace(Array.from(ungroupedComponents.values()))
    }

    return layers
  }

  private collectImportedFilenames (component: Component, components: Components, filenames: Set<string>) {
    if (filenames.has(component.filename)) return

    filenames.add(component.filename)

    if (!component.last) {
      component.imports.forEach(importedFilename => {
        const importedComponent = components.get(importedFilename)
        if (importedComponent) {
          this.collectImportedFilenames(importedComponent, components, filenames)
        }
      })
    } else {
      component.imports = []
    }
  }

  protected resolveConflictingComponentNames (components: Components): Components {
    const componentsByName: { [name: string]: Component[] } = {}

    for (const component of components.values()) {
      componentsByName[component.name] = componentsByName[component.name] || []
      componentsByName[component.name].push(component)
    }

    for (const name in componentsByName) {
      const components = componentsByName[name]
      const differentFilenames = new Set(
        components.map(component => component.filename)
      )
      const shouldPrefixWithDirectory = differentFilenames.size > 1 || name === 'index'

      if (shouldPrefixWithDirectory) {
        for (const component of components) {
          const dir = path.basename(path.dirname(component.filename))
          component.name = path.join(dir, component.name)
        }
      }
    }

    return components
  }

  protected sortComponentsByName (components: Components): Components {
    const sortedComponents: Components = new Map(Array.from(components.entries()).sort(
      (a, b) => a[1].name.localeCompare(b[1].name)
    ))

    for (const component of components.values()) {
      component.imports = component.imports
        .filter(importedFilename => components.has(importedFilename))
        .sort((a, b) => {
          const componentA = components.get(a)!
          const componentB = components.get(b)!
          return componentA.name.localeCompare(componentB.name)
        })
    }

    return sortedComponents
  }

  protected findComponentSchema (output: OutputSchema, filename: string): ComponentSchema {
    const componentSchema = this.config.components.find(componentSchema => {
      const outputFilters: ComponentFilters[] = [output, ...this.config.array(output.groups) || []]
      const includedInOutput = outputFilters.some(
        outputFilter => this.verifyComponentFilters(outputFilter, componentSchema)
      )

      if (includedInOutput) {
        return !!componentSchema.patterns && nanomatch.some(filename, componentSchema.patterns)
      } else {
        return false
      }
    })

    if (!componentSchema) {
      throw new Error(`Component schema not found: ${filename}`)
    }

    return componentSchema
  }

  protected verifyComponentFilters (filters: ComponentFilters, component: Component | ComponentSchema): boolean {
    const matchesPatterns = !('filename' in component) || !filters.patterns || nanomatch.some(component.filename, filters.patterns)
    const matchesComponents = !filters.components || filters.components.some(type => type === component.type)
    return matchesPatterns && matchesComponents
  }

  protected getComponentName (filename: string, componentConfig: ComponentSchema): string {
    const nameFormat = componentConfig.format

    if (nameFormat === ComponentNameFormat.FULL_NAME) {
      return path.basename(filename)
    }

    return path.basename(filename, path.extname(filename))
  }

  protected getAllComponents (layers: Layers, sortByName = false): Component[] {
    const components = ([] as Component[]).concat(...[...layers.values()].map(components => [...components]))

    if (sortByName) {
      components.sort((a, b) => a.name.localeCompare(b.name))
    }

    return components
  }
}
