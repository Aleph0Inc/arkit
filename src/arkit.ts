import * as fs from 'fs'
import * as path from 'path'
import { debug, trace } from './logger'
import { Config } from './config'
import { Parser } from './parser'
import { Generator } from './generator'

export interface Options {
  directory: string,
  output?: string[],
  first?: string[]
}

const getOptions = (options?: Options): Options => {
  const opts: Options = {
    directory: (options && options.directory) || '',
    output: (options && options.output) || [],
    first: (options && options.first) || []
  }
  const directory = path.isAbsolute(opts.directory) ? opts.directory : path.join(process.cwd(), opts.directory)

  return {
    ...opts,
    directory
  }
}

export const arkit = (options?: Options): Promise<string[]> => {
  const opts = getOptions(options)
  debug('Options')
  debug(opts)

  const config = new Config(opts)

  debug('Config')
  debug(config)

  const parser = new Parser(config)
  const files = parser.parse()

  trace('Parsed files')
  trace(files)

  const generator = new Generator(config, files)

  return Promise.all(config.outputs.reduce((promises, output) => {
    const puml = generator.generatePlantUML(output)

    if (output.path && output.path.length) {
      for (const outputPath of config.array(output.path)!) {
        promises.push(generator.convert(outputPath, puml))
      }
    } else {
      promises.push(generator.convert('svg', puml))
    }

    return promises
  }, [] as Promise<string>[]))
}
