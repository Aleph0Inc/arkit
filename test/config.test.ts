import * as path from 'path'
import { Config } from '../src/config'

describe('Config', () => {
  test('Sample', () => {
    const dir = path.join(__dirname, './sample')
    const config = new Config(dir)

    expect(config.directory).toBe(dir)
    delete config.directory
    expect(config).toMatchSnapshot()
  })

  test('Angular2 Todo', () => {
    const dir = path.join(__dirname, './angular2_es2015')
    const config = new Config(dir)

    expect(config.directory).toBe(dir)
    delete config.directory
    expect(config).toMatchSnapshot()
  })

  test('Express', () => {
    const dir = path.join(__dirname, './express')
    const config = new Config(dir)

    expect(config.directory).toBe(dir)
    delete config.directory
    expect(config).toMatchSnapshot()
  })
})
