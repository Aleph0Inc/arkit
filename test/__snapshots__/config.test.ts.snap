// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Angular2 Todo 1`] = `
Config {
  "components": Array [
    Object {
      "patterns": Array [
        "**/*.ts",
        "**/*.js",
        "**/*.jsx",
        "**/*.tsx",
      ],
      "type": "Component",
    },
  ],
  "excludePatterns": Array [
    "node_modules/**",
    "test/**",
    "**/*.test.*",
    "**/*.spec.*",
  ],
  "extensions": Array [
    ".js",
    ".ts",
    ".jsx",
    ".tsx",
  ],
  "outputs": Array [
    Object {},
  ],
  "patterns": Array [
    "**/*.ts",
    "**/*.js",
    "**/*.jsx",
    "**/*.tsx",
  ],
}
`;

exports[`Config Express 1`] = `
Config {
  "components": Array [
    Object {
      "patterns": Array [
        "**/*.ts",
        "**/*.js",
        "**/*.jsx",
        "**/*.tsx",
      ],
      "type": "Component",
    },
  ],
  "excludePatterns": Array [],
  "extensions": Array [
    ".js",
    ".ts",
    ".jsx",
    ".tsx",
  ],
  "outputs": Array [
    Object {
      "path": Array [
        "express.svg",
      ],
    },
  ],
  "patterns": Array [
    "**/*.ts",
    "**/*.js",
    "**/*.jsx",
    "**/*.tsx",
  ],
}
`;

exports[`Config Sample 1`] = `
Config {
  "components": Array [
    Object {
      "format": "full",
      "patterns": Array [
        "./main.ts",
      ],
      "type": "App",
    },
    Object {
      "patterns": Array [
        "**/*.view.**",
      ],
      "type": "View",
    },
    Object {
      "patterns": Array [
        "**/**component.**",
      ],
      "type": "Component",
    },
    Object {
      "patterns": Array [
        "**/controllers/*",
      ],
      "type": "Controller",
    },
    Object {
      "patterns": Array [
        "**/models/*",
      ],
      "type": "Model",
    },
  ],
  "excludePatterns": Array [],
  "extensions": Array [
    ".js",
    ".ts",
    ".jsx",
    ".tsx",
  ],
  "outputs": Array [
    Object {
      "groups": Array [
        Object {
          "components": Array [
            "View",
            "Controller",
          ],
          "type": "Main",
        },
        Object {
          "components": Array [
            "Component",
            "Model",
          ],
          "type": "Base",
        },
        Object {
          "patterns": Array [
            "**/*",
          ],
        },
      ],
      "path": Array [
        "./docs/architecture.puml",
      ],
    },
    Object {
      "path": "./docs/architecture.svg",
    },
  ],
  "patterns": Array [
    "./main.ts",
    "**/*.view.**",
    "**/**component.**",
    "**/controllers/*",
    "**/models/*",
  ],
}
`;
